### PR TITLE
PT-3181: Enable gifts in checkout

### PR DESCRIPTION
**Pull request by eMazeika**

### CODE REVIEW

File: src/XPurchase/VirtoCommerce.XPurchase/Commands/AddCartItemCommand.cs [hunk 1]
```diff
@@ -8,12 +8,13 @@ public AddCartItemCommand()
         {
         }
 
-        public AddCartItemCommand(string storeId, string cartType, string cartName, string userId, string currencyCode, string cultureName, string productId, int quantity, decimal? price, string comment)
+        public AddCartItemCommand(string storeId, string cartType, string cartName, string userId, string currencyCode, string cultureName, string productId, int quantity, decimal? price, string comment, bool isGift)
```
Comment (Vectorfield4) [inline src/XPurchase/VirtoCommerce.XPurchase/Commands/AddCartItemCommand.cs:11]: Breaking change. Undo this change. Use public setter.

File: src/XPurchase/VirtoCommerce.XPurchase/Extensions/RewardExtensions.cs [hunk 2]
```diff
@@ -52,6 +52,15 @@ public static void ApplyRewards(this ShoppingCart shoppingCart, ICollection<Prom
                 shoppingCart.Discounts.Add(discount);
                 shoppingCart.DiscountAmount += discount.DiscountAmount;
             }
+
+            // remove the (added) gifts, if corresponding reward is missing
+            foreach (var lineItem in (shoppingCart.Items ?? Enumerable.Empty<LineItem>()).Where(li => li.IsGift).ToArray())
```
Comment (Vectorfield4) [inline src/XPurchase/VirtoCommerce.XPurchase/Extensions/RewardExtensions.cs:57]:
```suggestion
            foreach (var lineItem in shoppingCart?.Items.Where(li => li.IsGift) ?? Enumerable.Empty<LineItem>())
```

File: src/XPurchase/VirtoCommerce.XPurchase/Extensions/RewardExtensions.cs [hunk 3]
```diff
@@ -52,6 +52,15 @@ public static void ApplyRewards(this ShoppingCart shoppingCart, ICollection<Prom
...
+            {
+                if (!rewards.OfType<GiftReward>().Any(re => re.ProductId == lineItem.ProductId))
```
Comment (Vectorfield4) [inline src/XPurchase/VirtoCommerce.XPurchase/Extensions/RewardExtensions.cs:59]: Add comment, please.

**Review by Vectorfield4 — CHANGES_REQUESTED** [review]

**Review by ksavosteev — COMMENTED** [review]
I suggest adding Gift lineitems to Items collection gift immediately after promo evaluation and saving them with IsGift flag, not by a mutation. Later we can filter them by the flag. 
We can have another collection for select gifts.

File: src/XPurchase/VirtoCommerce.XPurchase/Extensions/RewardExtensions.cs (hunk 3 above)
Comment (eMazeika) [reply src/XPurchase/VirtoCommerce.XPurchase/Extensions/RewardExtensions.cs:59]: yes, "// remove the (added) gifts, if corresponding reward is missing". Anything useful to add?

**Review by tatarincev — CHANGES_REQUESTED** [review]
Please see my notes about requirements and implementation details
https://virtocommerce.atlassian.net/browse/PT-3179
